### ISSUE TITLE
SSO/OAuth2/OIDC support

### ISSUE THREAD

**Issue by davispuh** [lobsters/lobsters#1408]
Currently only way to login is with username/password but I want to host my own instance where people can login using OpenID Connect (OIDC) instead of needing another password.
If I were to implement such feature would it be accepted/merged?
It would be enabled only if it's specifically configured so it wouldn't change anything for existing instances that don't want to use it.

I think most quickest way to implement this with least work would be to handle it at Nginx layer and simply pass `REMOTE_USER`/`REMOTE_EMAIL` headers to Lobsters.

**Comment by pushcx** [comment]
Yes, I’d merge and maintain it if we can keep it a well-contained optional feature so maintenance is cheap. Like we have the settings for `open_signups?` and `invitation_requests` even though Lobsters doesn’t use them.

I’d have to think a bit about whether we should. #342 and (especially) #1395 have made it real clear that it’s important to think through the full lifecycle of these features and their failure modes, especially when other parties may be unreliable.

I’d also love to hear about the site you’re setting up, feel free to email me. Grab me [on IRC](https://lobste.rs/chat) or in (office hours)[https://push.cx/stream) if you’d like to talk through anything in the codebase or your feature.

**Comment by davispuh** [comment]
Awesome!

> I’d have to think a bit about whether we should.

Think about what?

> I’d also love to hear about the site you’re setting up, feel free to email me.

It's very early stage on this project but basically idea is to create Q&A site (as an alternative to StackOverflow/Stack Exchange) that's specifically built for crowdsourcing data for training AIs. All content will be available under [CC0](https://creativecommons.org/public-domain/cc0/) and easily downloadable for any purpose. Currently big tech basically uses your data but puts restrictions (if allows use at all) if you want to use it aswell.
And instead of building it from scratch I think reusing Lobsters and building other software around could make it work with significantly less effort. And sure I'll contact you later once we've made more development on this.

**Comment by pushcx** [comment]
Sorry, I wrote my comment in a hurry, that “whether we should” didn’t make a lot of sense. I was trying to say I’ll be thinking about whether the Lobsters site itself should start using the feature.

Good luck with your site. As I said, happy to chat about the codebase.

**Comment by davispuh** [comment]
Hey, I decided that for Q&A site we'll be using [Apache Asnwer](https://answer.apache.org/) instead of Lobster as that seems to suit better.
But we still have a need for Lobster for different functionality and now I started thinking about this again and I realized putting Lobster behind auth Proxy won't work because I want that it's still readable without login (eg. search engines etc). Adding auth proxy would work if you always want authentication for everything.

So to support this functionality will need to implement proper OIDC support and have to use gem like `openid_connect` (or some other but this seems fine)
Essentially it's quite big change. Also for my use case I only need support for single OIDC provider but if you want to allow login thru Google/GitHub and other providers then you would need multiple provider support. Also in my case I don't want to allow password based login/registration but only OIDC.
